lib/router/src/cache.rs

```diff
@@ -1,426 +1,367 @@
-        let mut clock = Clock::default();
-        let mut cache =
-            Cache::<_, MultiplyAndAssign>::new(1, Duration::from_secs(0)).with_clock(clock.clone());
-
-        let t0 = clock.now();
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(333, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        let t1 = clock.now();
-        assert_eq!(cache.access(&333).map(|n| n.last_access()), Some(t0));
-
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(cache.access(&333).map(|n| n.last_access()), Some(t1));
-    }
+    fn store_and_background_purge() {
+        tokio::run(futures::lazy(|| {
+            let (cache, cache_purge) = Cache::new(2, Duration::from_millis(10));
 
-    #[test]
-    fn last_access_wiped_on_evict() {
-        let mut clock = Clock::default();
-        let mut cache =
-            Cache::<_, MultiplyAndAssign>::new(1, Duration::from_secs(0)).with_clock(clock.clone());
-
-        let t0 = clock.now();
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(333, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(cache.access(&333).map(|n| n.last_access()), Some(t0));
-
-        // Cause the router to evict the `333` route.
-        clock.advance(Duration::from_secs(1));
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(444, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        let t1 = clock.now();
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(333, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(cache.access(&333).map(|n| n.last_access()), Some(t1));
+            // Spawn a background purge
+            tokio::spawn(cache_purge);
+
+            {
+                let mut cache = cache.lock().unwrap();
+
+                cache.reserve().expect("reserve").store(1, 2);
+                cache.reserve().expect("reserve").store(2, 3);
+                assert_eq!(cache.vals.len(), 2);
+            }
+
+            // Sleep for an amount of time greater than expiry duration so
+            // that the background purge purges the entire cache.
+            tokio::spawn(
+                tokio_timer::sleep(Duration::from_millis(100))
+                    .map(move |_| {
+                        assert_eq!(cache.lock().unwrap().vals.len(), 0);
+                    })
+                    .map_err(|_| ()),
+            );
+
+            Ok(())
+        }))
     }
 
     #[test]
-    fn node_access_updated_on_drop() {
-        let mut clock = Clock::default();
-        let t0 = clock.now();
-        let mut node = Node::new(123, t0);
+    fn drop_access_resets_expiration() {
+        use tokio::runtime::current_thread::Runtime;
+
+        let mut runtime = Runtime::new().unwrap();
```

Comment:
TIOLI, but I might recommend rewriting the tests above that use `tokio::run`/`tokio::spawn` to also use the current_thread `Runtime`, for consistency? Also, the pattern of `block_on` -> assert -> `block_on` rather than chaining together futures that make assertions might lead to less "rightward drift" in indentation...